### PR TITLE
[ART-1343] Add multi-arch support on OLM Operators Metadata

### PR DESCRIPTION
**Pull request by sosiouxme**
Going to start a PR just so I can easily comment.
@thiagoalessio 

----
[ART-1343](https://jira.coreos.com/browse/ART-1343) && [ART-839](https://jira.coreos.com/browse/ART-839): Support 2 ways of building OLM operator metadata: `by-arch` or `manifest-list`.

Either by CLI option:
```
doozer -g openshift-4.2 \
  operator-metadata:build \
  cluster-logging-operator-container-v4.2.9-201911261133 \
  --stream dev \
  --image-ref-mode [by-arch|manifest-list]
```

Or by ocp-build-data's group.yml option:
* https://github.com/openshift/ocp-build-data/pull/241
* https://github.com/openshift/ocp-build-data/pull/242
* https://github.com/openshift/ocp-build-data/pull/243
* https://github.com/openshift/ocp-build-data/pull/244

## `by-arch` mode

when mode is `by-arch`, it creates additional manifests dirs and channels per arch, for example:
```
[cluster-logging-dev-operator-metadata]$ tree
.
├── Dockerfile
└── manifests
    ├── 4.1
    │   ├── cluster-loggings.crd.yaml
    │   ├── cluster-logging.v4.1.0.clusterserviceversion.yaml
    │   └── image-references
    ├── 4.2
    │   ├── cluster-loggings.crd.yaml
    │   ├── cluster-logging.v4.2.0.clusterserviceversion.yaml
    │   └── image-references
    ├── 4.2-s390x
    │   ├── cluster-loggings.crd.yaml
    │   ├── cluster-logging.v4.2.0-s390x.clusterserviceversion.yaml
    │   └── image-references
    ├── 4.3
    │   ├── cluster-loggings.crd.yaml
    │   ├── cluster-logging.v4.3.0.clusterserviceversion.yaml
    │   ├── collectors.crd.yaml
    │   ├── image-references
    │   └── logforwardings.crd.yaml
    └── cluster-logging.package.yaml
```
And image references will point to the `SHA` of a build of that image for the specific architecture.
It also lists each additional architecture as another **channel** in `package.yml`:
```
channels:
- currentCSV: clusterlogging.4.2.9-201911261133
  name: '4.2'

- currentCSV: clusterlogging.4.1.27-201912030019
  name: preview

- currentCSV: clusterlogging.4.3.0-201912020833
  name: '4.3'

- currentCSV: clusterlogging.4.2.9-201911261133-s390x
  name: 4.2-s390x

defaultChannel: '4.3'
packageName: cluster-logging
```

## `manifest-list` mode

When mode is `manifest-list`, no additional dirs are created and image references will point to the `SHA` of a Digest, containing `SHAs` for every architecture that the image was built.

## Still TODO

* [ ] Write more tests to demonstrate both ways of building metadata repo

### CODE REVIEW

File: doozerlib/operator_metadata.py [hunk 1]
```diff
@@ -52,7 +52,10 @@ def update_and_build(nvr, stream, runtime, merge_branch, force_build=False):
     :param string merge_branch: Which branch should be updated in the metadata repo
     :return bool True if operations succeeded, False if something went wrong
     """
-    op_md = OperatorMetadataBuilder(nvr, stream, runtime=runtime)
+    if runtime.group_config.vars['MAJOR'] == 4 and runtime.group_config.vars['MINOR'] == 2:
```
Comment (sosiouxme) [inline doozerlib/operator_metadata.py]: really dislike anything version-specific in code. can we just extend the `operator_ref_arch` implementation from https://jira.coreos.com/browse/ART-839 and configure it with
```
operator_ref_arch: multichannel
```
(and take the enabled arch list from group.yml)

File: doozerlib/operator_metadata.py [hunk 2]
```diff
@@ -179,7 +182,7 @@ def update_current_csv_shasums(self):
             with open(file, 'w') as writer:
                 writer.write(self.replace_version_by_sha_on_image_references(contents))
 
-    def replace_version_by_sha_on_image_references(self, contents):
+    def replace_version_by_sha_on_image_references(self, contents, arch='amd64'):
```
Comment (sosiouxme) [inline doozerlib/operator_metadata.py]: please, `x86_64` unless we're in golang world.

**Comment by thiagoalessio** [comment]
:+1: @sosiouxme ! just "made it work" ... now i'll start "making it right" ;D

**Comment by thiagoalessio** [comment]
just as a reference, this is how the manifests look like now:

http://pkgs.devel.redhat.com/cgit/containers/cluster-logging-dev-operator-metadata/tree/manifests?h=rhaos-4-rhel-7&id=a9f9212075ff411233ca8c6210b2f3661a3035db

https://brewweb.engineering.redhat.com/brew/buildinfo?buildID=1027510

**Comment by thiagoalessio** [comment]
Removed **WIP**, even though there are more tests to be written, production code could be merged if needed.

**Comment by sosiouxme** [comment]
:shipit:

**Pull request merged**